### PR TITLE
feat: add priorityclassname and lifecycle

### PR DESCRIPTION
**Pull request by rtest12**
add priorityclassname and lifecycle

Screenshot from 2024-06-03 14-55-37
Screenshot from 2024-05-29 17-55-36
Untitled

### CODE REVIEW

**Comment by rtest12** [comment]
@AsfaMumtaz @rasheedamir 
ptl

**Comment by rasheedamir** [comment]
@d3adb5 can you review it?

File: application/templates/deployment.yaml [hunk 1]
```diff
@@ -141,6 +144,10 @@ spec:
         image: {{ $image }}
 
         imagePullPolicy: {{ .Values.deployment.image.pullPolicy }}
+        {{- if .Values.deployment.lifecycle }}
+        lifecycle:
+{{ toYaml .Values.deployment.lifecycle | indent 10 }}
```
Comment (d3adb5) [inline application/templates/deployment.yaml:149]: Let's try to keep indentation levels consistent with where the templated values will end up.

```suggestion
        lifecycle:
          {{- toYaml .Values.deployment.lifecycle | nindent 10 }}
```

**Review by d3adb5 — APPROVED** [review]
I left one suggestion, and otherwise this LGTM, however I want to note that the majority of these changes have to do with whitespace and formatting. While that kind of tidying up is welcome, I'd much prefer it to happen in its own commit. Can you perhaps separate them, @rtest12?

Could you chime in here, @rasheedamir?

**Comment by rtest12** [comment]
> I left one suggestion, and otherwise this LGTM, however I want to note that the majority of these changes have to do with whitespace and formatting. While that kind of tidying up is welcome, I'd much prefer it to happen in its own commit. Can you perhaps separate them, @rtest12?
> 

@rasheedamir @d3adb5 
Ok done, I've split the logic into two separate commits.

**Review by d3adb5 — DISMISSED** [review]
LGTM! Thank you, @rtest12.

**Comment by rtest12** [comment]
@rasheedamir Could you take a look and release this PR please?

**Comment by rtest12** [comment]
Any news?
@d3adb5 @hanzala1234 @mustafaStakater @usamaahmadkhan @kahootali @rasheedamir  @aslafy-z @AsfaMumtaz @sabkat @KhizerJaan
Could you take a look and release this PR please?

File: README.md [hunk 2]
```diff
@@ -54,25 +54,27 @@ To uninstall the chart:
 | deployment.revisionHistoryLimit          | The number of old history to retain to allow rollback                                                                                      | `2`            |
 | deployment.command                       | Command for primary container of deployment                                                                                                | `[]`           |
 | deployment.args                          | Arg for primary container of deployment                                                                                                    | `[]`           |
+| deployment.priorityClassName             | Defines the priority class for pod scheduling                                                                                              | `""`           |
```
Comment (aslafy-z) [inline README.md:57]: `priorityClassName` is missing in the CronJob section of the documentation.

File: application/templates/deployment.yaml [hunk 3]
```diff
@@ -75,19 +75,22 @@ spec:
...
       - name: {{ $key }}
 {{ include "application.tplvalues.render" ( dict "value" $value "context" $ ) | indent 8 }}
       {{- end }}
       {{- end }}
       {{- if .Values.deployment.nodeSelector }}
       nodeSelector:
 {{ toYaml .Values.deployment.nodeSelector | indent 8 }}
+      {{- end }}
+      {{- if .Values.deployment.priorityClassName }}
+      priorityClassName: {{ .Values.deployment.priorityClassName }}
```
Comment (aslafy-z) [inline application/templates/deployment.yaml:93]: Prefer to use `with` block if possible

File: application/templates/deployment.yaml [hunk 4]
```diff
@@ -141,6 +144,10 @@ spec:
         image: {{ $image }}
 
         imagePullPolicy: {{ .Values.deployment.image.pullPolicy }}
+        {{- if .Values.deployment.lifecycle }}
```
Comment (aslafy-z) [inline application/templates/deployment.yaml:147]: Prefer to use `with` block if possible

File: application/values.yaml [hunk 5]
```diff
@@ -73,17 +74,17 @@ deployment:
 
   # Init containers which runs before the app container
   initContainers:
-#      init-contaner:
-#        image: busybox
-#        imagePullPolicy: IfNotPresent
-#        command: ['/bin/sh']
+    #  init-contaner:
```
Comment (aslafy-z) [inline application/values.yaml:77]: This change is odd, alignment is wrong compared to some other changes. Can you please align correctly?

**Review by aslafy-z — COMMENTED** [review]
Hey, thank you for your contrib. I added a few comments.

File: README.md (hunk 2 above)
Comment (rtest12) [reply README.md:57]: @aslafy-z 
right, done

File: application/templates/deployment.yaml (hunk 3 above)
Comment (rtest12) [reply application/templates/deployment.yaml]: @aslafy-z 
done

File: application/templates/deployment.yaml (hunk 4 above)
Comment (rtest12) [reply application/templates/deployment.yaml]: @aslafy-z 
done

File: application/values.yaml (hunk 5 above)
Comment (rtest12) [reply application/values.yaml]: @aslafy-z 
Sure, done

**Comment by rtest12** [comment]
> Hey, thank you for your contrib. I added a few comments.

@aslafy-z @d3adb5 
Done, please re-approve

File: application/templates/deployment.yaml [hunk 6]
```diff
@@ -141,6 +144,10 @@ spec:
         image: {{ $image }}
 
         imagePullPolicy: {{ .Values.deployment.image.pullPolicy }}
+        {{- with .Values.deployment.lifecycle }}
+        lifecycle:
+          {{- toYaml .Values.deployment.lifecycle | nindent 10 }}
```
Comment (aslafy-z) [inline application/templates/deployment.yaml:149]:
```suggestion
          {{- toYaml . | nindent 10 }}
```

**Review by aslafy-z — CHANGES_REQUESTED** [review]
Missing change that break ci

**Comment by rtest12** [comment]
> Missing change that break ci

@aslafy-z 
Done

**Comment by aslafy-z** [comment]
This PR now hits the same issue as https://github.com/stakater/application/pull/320#pullrequestreview-2161390342.
Will ping stakater team so they fix it asap. 
I'll approve and merge once the build passes.

**Comment by rtest12** [comment]
> This PR now hits the same issue as [#320 (review)](https://github.com/stakater/application/pull/320#pullrequestreview-2161390342), reported as #321. Will ping stakater team so they fix it asap. I'll approve and merge once the build passes.

@aslafy-z 
Well thank you

**Comment by rtest12** [comment]
> @rtest12 CI is OK, can you please rebase your PR?

@aslafy-z 
Done, please check again

**Comment by aslafy-z** [comment]
@rtest12 Helm lint fails with an error. Can you give a look please?

**Comment by rtest12** [comment]
> @rtest12 Helm lint fails with an error (you can safely ignore the warnings). Can you give a look please?

@aslafy-z Could you attach a screenshot?

File: application/templates/deployment.yaml (hunk 6 above)
Comment (aslafy-z) [reply application/templates/deployment.yaml:149]: image

I think this is still to be applied. @rtest12

**Comment by aslafy-z** [comment]
See https://github.com/stakater/application/pull/313#discussion_r1671828812

**Comment by rtest12** [comment]
> @rtest12 See [#313 (comment)](https://github.com/stakater/application/pull/313#discussion_r1671828812)

@aslafy-z 
Sure, done

**Comment by rasheedamir** [comment]
@d3adb5 can you approve it as well if all good? so, it can be merged?

**Review by d3adb5 — APPROVED** [review]
Looks good to me! Thanks, @rtest12.

**Pull request merged**